File: python/problems/leetcode/0009_palindrome_number/solution.py
"""LeetCode 9: Palindrome Number."""

from __future__ import annotations


class Solution:
    def isPalindrome(self, x: int) -> bool:
        if x < 0 or (x % 10 == 0 and x != 0):
            return False
        rev = 0
        while x > rev:
            rev = rev * 10 + x % 10
            x //= 10
        return x == rev or x == rev // 10


if __name__ == "__main__":
    assert Solution().isPalindrome(121)
    assert not Solution().isPalindrome(-121)
    assert Solution().isPalindrome(0)
    assert not Solution().isPalindrome(10)
    print("0009 OK")

File: cpp/problems/leetcode/0009_palindrome_number/solution.cpp
// LeetCode 9
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isPalindrome(int x) {
        if (x < 0 || (x % 10 == 0 && x != 0)) return false;
        long long rev = 0;
        int xx = x;
        while (xx > rev) {
            rev = rev * 10 + xx % 10;
            xx /= 10;
        }
        return xx == rev || xx == rev / 10;
    }
};

int main() {
    assert(Solution().isPalindrome(121));
    assert(!Solution().isPalindrome(-121));
    assert(Solution().isPalindrome(0));
    assert(!Solution().isPalindrome(10));
    cout << "0009 OK" << endl;
    return 0;
}
